Print `IMG` auto-sizes contain CSS fix by enqueueing inline style

westonruter opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/62731

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

Instead of manually printing the `STYLE` tag to fix `sizes=auto` issue with images, this enqueues the inline style via `WP_Styles`. This ensures the styles are printed with other styles as opposed to being printed very early, even before the `TITLE` tag. This also allows plugins to do additional processing of the stylesheet prior to it being printed. Note that `array_unshift( wp_scripts()->queue, $handle )` is used instead of `wp_enqueue_style( $handle )` in order to ensure that the stylesheet is printed before any others, preserving the existing cascade.

This is very similar to the change that was made to replace `print_emoji_styles()` with `wp_enqueue_emoji_styles()` in Core-58775.

- [x] Determine whether [extreme back-compat is warranted](https://core.trac.wordpress.org/ticket/62731#comment:7).
- [x] Add tests.

<!-- Insert a description of your changes here -->

Diff on rendered page (after Prettier formatting):

```diff
--- /tmp/before.html	2025-06-10 16:25:19
+++ /tmp/after.html	2025-06-10 16:25:39
@@ -4,11 +4,6 @@
 		<meta charset="UTF-8" />
 		<meta name="viewport" content="width=device-width, initial-scale=1" />
 		<meta name="robots" content="max-image-preview:large" />
-		<style>
-			img:is([sizes='auto' i], [sizes^='auto,' i]) {
-				contain-intrinsic-size: 3000px 1500px;
-			}
-		</style>
 		<title>WordPress Develop</title>
 		<link
 			rel="alternate"
@@ -544,6 +539,11 @@
 				}
 			)(window, document, window._wpemojiSettings);
 		</script>
+		<style id="wp-img-auto-sizes-contain-inline-css">
+			img:is([sizes='auto' i], [sizes^='auto,' i]) {
+				contain-intrinsic-size: 3000px 1500px;
+			}
+			/*# sourceURL=wp-img-auto-sizes-contain-inline-css */
+		</style>
 		<style id="wp-block-site-title-inline-css">
 			.wp-block-site-title {
 				box-sizing: border-box;
```

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

westonruter committed:
Print IMG auto-sizes contain CSS fix by enqueueing inline style

westonruter committed:
Do enqueueing in existing function

westonruter commented on src/wp-includes/media.php:
```diff
@@ -2077,9 +2079,12 @@ function wp_print_auto_sizes_contain_css_fix() {
+	wp_register_style( $handle, false );
+	wp_add_inline_style( $handle, 'img:is([sizes="auto" i], [sizes^="auto," i]) { contain-intrinsic-size: 3000px 1500px }' );
+
+	// Make sure inline style is printed first.
```
```suggestion
	// Make sure inline style is printed first since it was previously printed at wp_head priority 1 and this preserves the CSS cascade.
```

joemcgill commented on src/wp-includes/media.php:
```diff
@@ -2077,9 +2079,12 @@ function wp_print_auto_sizes_contain_css_fix() {
+	wp_add_inline_style( $handle, 'img:is([sizes="auto" i], [sizes^="auto," i]) { contain-intrinsic-size: 3000px 1500px }' );
+
+	// Make sure inline style is printed first.
+	array_unshift( wp_styles()->queue, $handle );
```
This seems a bit hacky to me, but I understand the purpose. I assume there's not a better way to control the order of registered styles besides maybe making sure that the `wp_register_style()` call is made on an earlier hook, like `init`, correct?

westonruter replied:
Exactly.

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -2061,14 +2061,16 @@ function wp_sizes_attribute_includes_valid_auto( string $sizes_attr ): bool {
+ * @link https://html.spec.whatwg.org/multipage/rendering.html#img-contain-size
+ * @link https://core.trac.wordpress.org/ticket/62413
  */
 function wp_print_auto_sizes_contain_css_fix() {
```
I don't love this name now, since it's implying this function does something that it doesn't.

We could potentially deprecate (effectively rename) it, though let's discuss how reasonable that would be.

* Obviously there's a BC risk if people expect this function to be hooked into `wp_head` and want to remove it to disable the snippet.
* However, there's also a BC risk if we _just_ change what this function does, in case someone calls it manually somewhere else.
* If someone unhooks this function to disable the snippet, and we wanted to replace it with a new function, we could still add the hook first, see if it is removed, and only before the function would be called ensure it is not invoked. This way we know whether it was removed, and could "forward" this removal to the new function.
    * I believe we did something like that in Core somewhere else before, but I can't recall where exactly.
* A nice side-effect of deprecating / replacing would be that we could also use the proper hook `wp_enqueue_scripts` for the new function.

westonruter replied:
@felixarntz see https://github.com/WordPress/wordpress-develop/pull/8954/commits/e7144a3652f8888b1690db58acf361e4d0093912 for that which was the original approach I took.

> I believe we did something like that in Core somewhere else before, but I can't recall where exactly.

Yes. We did it with printing the emoji styles in Core-58775.

westonruter replied:
See also https://core.trac.wordpress.org/ticket/62731#comment:7

westonruter replied:
If we go with the function rename and move from `wp_head` to `wp_enqueue_scripts`, since ~this function was introduced with the `wp_img_tag_add_auto_sizes` filter to disable the style, it seems overkill to also retain back-compat to see if someone removed the action on `wp_head`. If they had done this, they were essentially doing it wrong.~ There are only [4 examples of this being done on GitHub](https://github.com/search?q=%2F%28has_action%7Cremove_action%29%5C%28%5Cs*%5B%27%22%5Dwp_head%5B%27%22%5D%2C%5Cs*%5B%27%22%5Dwp_print_auto_sizes_contain_css_fix%2F&type=code).

**Update:** The `wp_img_tag_add_auto_sizes` filter doesn't just prevent the style from being printed. It also omits `sizes=auto` from the `IMG` tags as well. See https://github.com/WordPress/wordpress-develop/pull/8954#issuecomment-3364076685

westonruter replied:
I've opened pull requests to use the filter instead:

1. https://github.com/szepeviktor/zero-bytes-theme/pull/3 (https://github.com/szepeviktor/zero-bytes-theme/pull/4)
1. https://github.com/yuupiyo/MyPortfolio/pull/1
1. https://github.com/pavelleonenko007/Utopia-wordpress-theme/pull/1
1. https://github.com/conedevelopment/base-starter-theme/pull/4

In any case, if someone doesn't update the hook, the risk of breakage is very low, as it will just mean an additional `STYLE` is output unexpectedly.

westonruter replied:
@felixarntz I believe your concerns are addressed by introducing a separate function here. If so, please approve the PR.

westonruter replied:
After re-examining the purpose of the `wp_img_tag_add_auto_sizes` filter, I realize I was wrong to suggest that it was the right way to prevent the style from being printed. See https://github.com/WordPress/wordpress-develop/pull/8954#issuecomment-3364076685

westonruter committed:
Revert "Do enqueueing in existing function"

This reverts commit e7144a3652f8888b1690db58acf361e4d0093912.

westonruter committed:
Add tests for wp_enqueue_img_auto_sizes_contain_css_fix

westonruter committed:
Add clarifying comment

westonruter committed:
fixup! Add tests for wp_enqueue_img_auto_sizes_contain_css_fix

westonruter commented on src/wp-includes/media.php:
```diff
@@ -2061,25 +2061,29 @@ function wp_sizes_attribute_includes_valid_auto( string $sizes_attr ): bool {
  * `width: auto` or `width: fit-content` to appear smaller.
  *
- * @since 6.7.1
+ * @since 6.9.0
```
This could be 6.8.2 if we want.

```suggestion
 * @since 6.8.2
```

westonruter commented on src/wp-includes/deprecated.php:
```diff
@@ -6460,4 +6460,31 @@ function wp_add_editor_classic_theme_styles( $editor_settings ) {
+ * `width: auto` or `width: fit-content` to appear smaller.
+ *
+ * @since 6.7.1
+ * @deprecated 6.9.0 Use wp_enqueue_img_auto_sizes_contain_css_fix() instead.
```
Per below, if we want to target the next minor release:

```suggestion
 * @deprecated 6.8.2 Use wp_enqueue_img_auto_sizes_contain_css_fix() instead.
```

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into fix/trac-62731

westonruter commented:
I just merged in the latest from `trunk`. Note there is now a `sourceURL` comment present:

```html
<style id='wp-img-auto-sizes-contain-inline-css'>
img:is([sizes="auto" i], [sizes^="auto," i]) { contain-intrinsic-size: 3000px 1500px }
/*# sourceURL=wp-img-auto-sizes-contain-inline-css */
</style>
```

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -2081,25 +2081,29 @@ function wp_sizes_attribute_includes_valid_auto( string $sizes_attr ): bool {
  * @see https://core.trac.wordpress.org/ticket/62413
  */
-function wp_print_auto_sizes_contain_css_fix() {
+function wp_enqueue_img_auto_sizes_contain_css_fix(): void {
```
Shouldn't we, either within this function, or in some other logic that runs on a hook shortly before this, check for `has_action( 'wp_head', 'wp_print_auto_sizes_contain_css_fix' )`, and only remove the deprecated action at the point that we know it wasn't _intentionally_ removed by any 3P code?

I think we should include this for BC, to make sure sites that removed the original hook don't see adverse effects now because we changed the hook. This is similar to how we did it elsewhere before I think.

westonruter replied:
@felixarntz Thanks for raising this again. After re-examining the purpose of the `wp_img_tag_add_auto_sizes` filter, I realize I was wrong to suggest that it was the right way to prevent the style from being printed. See https://github.com/WordPress/wordpress-develop/pull/8954#issuecomment-3364076685

westonruter committed:
Allow removal of original print action to prevent enqueueing

Co-authored-by: Felix Arntz <flixos90@git.wordpress.org>

westonruter committed:
Minify CSS

westonruter commented:
In 67d527c I've addressed the concerns with backwards-compatibility with themes/plugins that have intended to prevent this style from being printed by unhooking the action via:

```php
remove_action( 'wp_head', 'wp_print_auto_sizes_contain_css_fix', 1 );
```

In looking at this further, I realize I was mistaken in thinking that the `wp_img_tag_add_auto_sizes` was how this style was actually supposed to be omitted. In reality, this filter does much more, as it prevents `sizes=auto` from being added to all images as well. Therefore, I was wrong to suggest the proper way to simply omit this style was to do `add_filter( 'wp_img_tag_add_auto_sizes', '__return_false' )`.

In order to maintain backwards-compatibility, the now-deprecated `wp_print_auto_sizes_contain_css_fix()` function now remains hooked at `wp_head` with a priority 1, and now the `wp_enqueue_img_auto_sizes_contain_css_fix()` function is hooked at `wp_head` with priority 0, just before `wp_print_auto_sizes_contain_css_fix()` is slated to run. The first thing that `wp_print_auto_sizes_contain_css_fix()` does is it checks to see if `has_action( 'wp_head', 'wp_print_auto_sizes_contain_css_fix' )`. If not, then it short-circuits. Otherwise, it goes ahead and removes the action to prevent the deprecation notice, while also proceeding with enqueueing the CSS fix.

I've added tests for these various scenarios.



<!--
This is graveyard content which I abandoned sharing, but I didn't want to lose it just in case!

We did do it for `wp_enqueue_emoji_styles()`, but that function didn't have a dedicated filter specifically for short-circuiting the printing of the style, like we have in `wp_img_tag_add_auto_sizes`. (It was also done for `the_block_template_skip_link()`, at last according to the comments in `default-filters.php`.)

There were only four examples of hook function being removed in this way: https://github.com/WordPress/wordpress-develop/pull/8954#discussion_r2143431106 

One of those has been fixed. 

Well, now [there are eight](https://github.com/search?q=%2F%28has_action%7Cremove_action%29%5C%28%5Cs*%5B%27%22%5Dwp_head%5B%27%22%5D%2C%5Cs*%5B%27%22%5Dwp_print_auto_sizes_contain_css_fix%2F&type=code).

There are currently [78 files](https://github.com/search?q=%2F%28add_filter%29%5C%28%5Cs*%5B%27%22%5Dwp_img_tag_add_auto_sizes%5B%27%22%5D%2C%2F&type=code) that use the `wp_img_tag_add_auto_sizes` filter.

In any case, if someone doesn't update the hook, the risk of breakage is very low, as it will just mean an additional `STYLE` is output unexpectedly.

The back-compat hack we did for `wp_enqueue_emoji_styles()` is somewhat painful, as in unit tests it has often meant needing to add this code:

```php
remove_action( 'wp_print_styles', 'print_emoji_styles' );
```

We've had to do this in the PerfLab repo as well. It seems this is needed whenever a test calls `wp_print_styles()` without first being sure to call `wp_enqueue_scripts()`, as otherwise `wp_enqueue_emoji_styles()` isn't called, in which case `remove_action( 'wp_print_styles', 'print_emoji_styles' )` never happens.

----

If we did that same thing to have back-compat for themes and plugins that unhook `wp_print_auto_sizes_contain_css_fix()` from `wp_head`, then here's what I think that would require:




then we'd need to back back-compat for `wp_print_auto_sizes_contain_css_fix()` actually printing styles, since it is supposed to run at `wp_head` priority 1, whereas `wp_enqueue_scripts()` happens later at `. It would instead need to do:

```

```

then we'd need to revert this change to `default-filters.php`:

```diff
- add_action( 'wp_head', 'wp_print_auto_sizes_contain_css_fix', 1 ); 
+ add_action( 'wp_enqueue_scripts', 'wp_enqueue_img_auto_sizes_contain_css_fix' );
```

And instead do this:

```diff
  add_action( 'wp_head', 'wp_print_auto_sizes_contain_css_fix', 1 );
+ add_action( 'wp_enqueue_scripts', 'wp_enqueue_img_auto_sizes_contain_css_fix' );
```



And then we'd need to have a different back-compat break here by having `wp_print_auto_sizes_contain_css_fix()` to not actually print anything. Either we'd have it be a no-op function and then have `wp_enqueue_img_auto_sizes_contain_css_fix()` do:

```php
if ( ! has_action( 'wp_head', 'wp_print_auto_sizes_contain_css_fix' ) ) {
    return;
}
```

Or, we could have `default-filters.php` have this instead:

```diff
  add_action( 'wp_head', 'wp_print_auto_sizes_contain_css_fix', 1 );
+ add_action( 'wp_enqueue_scripts', 'wp_enqueue_img_auto_sizes_contain_css_fix' );
```

 instead we'd have `wp_print_auto_sizes_contain_css_fix()` do:

```php
add_action( 'wp_enqueue_scripts', 'wp_enqueue_img_auto_sizes_contain_css_fix' );
```

Since we have a proper `wp_img_tag_add_auto_sizes` filter for disabling this style from being printed in the first place, I think any other method being used to disable it should be considered `_doing_it_wrong()`.
-->

westonruter committed:
Make test_wp_enqueue_img_auto_sizes_contain_css_fix less brittle

felixarntz reviewed (approved):
@westonruter Thank you for the updates, looks good to go IMO!
